Add dependency graph functionality

teodutu opened the pull request:
Changes:
- SandScout Compiler can now return the rules it discovers as a dictionary of lists.
- When the script is imported, only the dictionary output is supported.
- When called directly, the script uses argparse to be more verbose about its parameters.
- Converted the script to Python3.
- Improved coding style consistency.
- Increased tab size to 4 spaces.
- TODOs still present from before.

teodutu committed:
Add dependency graph functionality

Changes:
- SandScout Compiler can now return the rules it discovers as a dictionary of lists.
- When the script is imported, only the dictionary output is supported.
- When called directly, the script uses argparse to be more verbose about its parameters.
- Converted the script to Python3
- Improved coding style consistency.
- Increased tab size to 4 spaces.
- TODOs still present from before.

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -1,4 +1,5 @@
-#!/usr/bin/env python
+from argparse import ArgumentParser
```
Either add `#!/usr/bin/env python` shebang or remove execution permissions of script.

razvand replied:
Add top-level module docstring documenting this program.

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -1,4 +1,5 @@
-#!/usr/bin/env python
+from argparse import ArgumentParser
+
```
Please run the script through `pycodestyle` and / or `pylint` and fix reasonable warnings.

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -12,24 +13,25 @@
 ####################################################
```
Add simple comments for each global variable and function. I recommend using [docstrings](https://realpython.com/python-comments-guide/#python-commenting-best-practices).

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -39,335 +41,417 @@
+_GRAPH = {}
+_CMD_ARGS = None
+
 # Regular expression rules for simple tokens
```
These expressions are not used in the program.

teodutu replied:
They are. `ply` requires a set of tokens that it tries to match against the input file. They are declared in the `tokens` list above

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -39,335 +41,417 @@
 
-lexer.input(inputFile)
 
 """
```
Remove unused / commented code.

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -39,335 +41,417 @@
+# TODO: this is sort of a hack and I should evaluate it effects carefully.
+# The issue is that requireNot can now accept entitlement requirements as
+# parameters.
 def p_simpleEntValObject(p):
```
Do not use CamelCase, it's inconsistent with the rest of the names.

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -39,335 +41,417 @@
+    p[0] = [p[0]]
 
 
 def p_reqEnt(p):
```
Do not use CamelCase, it's inconsistent with the rest of the names.

razvand commented on profile_compilation/sandscout_compiler.py:
```diff
@@ -39,335 +41,417 @@
+
+
+def main():
+    _get_args()
```
Why is this using and underscore prefix? Is there a Python naming convention for this?

teodutu replied:
Yes. They're somewhat similar to `static` functions in C. If someone were to `from sandscout_compiler import *` [1], symbols starting with `_` would not be visible. However, when importing the file like this `import sandscout_compiler`, all symbols, regardless of names, would be imported.

[1] https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

razvand reviewed (changes requested):
Thanks for doing this. Please address my comments that mostly deal with the old version of the code. But, as long as you worked on it, it makes sense to improve it further.

teodutu committed:
Fix coding style

Changes:
- Fix all pycodestyle errors and those pylint errors that don't affect the
functionality of the code
- Add docstrings to all functions and to the script itself
- Replace camelCase with snake_case

teodutu committed:
Add fixes for SandBlaster

The output of the SandScout compiler differed in certain aspects from
that of SandBlaster. Now those differences have been eliminated.

teodutu committed:
Fix require-entitlement bug
